internals/controllers/migrationhierarchy/migrationhierarchy_controller.go

```diff
@@ -0,0 +1,440 @@
+	newSNS, reason, er := r.createSNS(oldSNS, currentNamespace, toNamespace)
+	if er != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed creating subnamespace '%s' under namespace '%s': "+er.Error(), currentNamespace, toNamespace)
+	}
+	logger.Info("successfully created new subnamespace under new parent", "subnamespace", currentNamespace, "new parent", toNamespace)
+
+	// delete the old subnamespace under the original parent
+	if reason, er := r.deleteSNS(oldSNS); er != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+
+		// if deleting the old subnamespace fails, delete the newly created subnamespace
+		if reason, er := r.deleteSNS(newSNS); er != nil {
+			err := r.updateMHErrorStatus(mhObject, er, reason)
+			if err != nil {
+				return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+			}
+			return fmt.Errorf("failed deleting subnamespace '%s' :"+err.Error(), newSNS.GetName())
+		}
+		return fmt.Errorf("failed deleting subnamespace '%s' :"+er.Error(), oldSNS.GetName())
+	}
+	logger.Info("successfully deleted old subnamespace from old parent", "subnamesapce", currentNamespace, "old parent", sourceSNSParentName)
+
+	toNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: toNamespace}, &corev1.Namespace{})
+	if err != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), toNamespace)
+	}
+
+	if err := r.UpdateNSBasedOnParent(ctx, toNS, ns); err != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed updating the labels and annotations of namespace '%s' according to its parent '%s': "+err.Error(), ns.GetName(), toNS.GetName())
+	}
+	logger.Info("successfully updated namespace based on parent", "subnamesapce", currentNamespace, "parent", toNamespace)
+
+	if err := r.UpdateAllNSChildrenOfNs(ctx, ns); err != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed updating labels and annotations of child namespaces of sunamespace '%s': "+err.Error(), ns.GetName())
+	}
+	logger.Info("successfully updated labels and annotations children namespaces of subnamespace", "subnamespace", currentNamespace)
+
+	if err := r.updateRole(toNS, danav1.NoRole); err != nil {
```

Comment:
wrap in function

Reply:
included it inside the update related objects function